CVE dependencies and github actions

nmccready opened the pull request:
- **fix: bump dependencies to remove CVE issues and modernice eslint / mocha**
- **chore: githhub actions for automations**

nmccready committed:
fix: bump dependencies to remove CVE issues and modernice eslint / mocha

nmccready committed:
chore: githhub actions for automations

nmccready committed:
fix: eslint-config-gulp

phated commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,19 @@
+version: 2
+updates:
+  - package-ecosystem: "npm"
+    directory: "/" # Location of package manifests
```
I'd recommend using
```
cooldown: 
  default-days: 7
```
As recommended by https://blog.yossarian.net/2025/11/21/We-should-all-be-using-dependency-cooldowns

phated commented on .github/workflows/_publish.yml_disable:
```diff
@@ -0,0 +1,29 @@
+    steps:
+      - uses: actions/checkout@v4
+      - name: Use Node.js ${{ matrix.node-version }}
+        uses: actions/setup-node@v6
```
Recommend pinning actions to commit hashes as per the default behavior of "unpinned hashes" in zizmor: https://docs.zizmor.sh/audits/#unpinned-uses

nmccready replied:
Need more context? Suggesting a full version pin here? Sorry I have not got into linting Actions yet... yay more boilerplate .

nmccready-tars replied:
Done! Pinned all GitHub Actions to their commit hashes across all workflow files:

- `actions/checkout@v4` → `actions/checkout@34e114876b0b11c390a56381ad16ebd13914f8d5`
- `actions/setup-node@v3` → `actions/setup-node@3235b876344d2a9aa001b8d1453c930bba69e610`
- `actions/setup-node@v6` → `actions/setup-node@6044e13b5dc448c55e2357c09f80417699197238`

Version tags are preserved in trailing comments for readability. This follows the zizmor 'unpinned-uses' security recommendation.

phated commented on .github/workflows/_publish.yml_disable:
```diff
@@ -0,0 +1,29 @@
+        uses: actions/setup-node@v6
+        with:
+          node-version: '20.x'
+          registry-url: 'https://registry.npmjs.org'
```
```suggestion
          registry-url: 'https://registry.npmjs.org
```
typo

nmccready replied:
Were you meaning to get rid of both quotes?

phated commented on .github/workflows/_publish.yml_disable:
```diff
@@ -0,0 +1,29 @@
+          registry-url: 'https://registry.npmjs.org'
+      - name: Publish to npm
+        run: | # npm 11.15.1 for OIDC support
+          npm install -g npm@11
```
perhaps use a newer node to publish to avoid needing to install an unpinned npm version

nmccready replied:
Yeah, currently still running with mainly 20 for most things and older npm, only recently moved to 11 NPM for OIDC support.

phated commented on .github/workflows/_publish.yml_disable:
```diff
@@ -0,0 +1,29 @@
+        run: | # npm 11.15.1 for OIDC support
+          npm install -g npm@11
+          npm ci
+          npm install
```
Don't bare `npm install` if you are including a lockfile. Always `npm ci`

nmccready-tars replied:
Fixed! Changed all `npm install` to `npm ci` in tests.yml and release.yml, and enabled npm caching since we now have a lockfile. Also removed the unnecessary `npm install -g npm@11` from tests (only needed for OIDC publish).

phated commented on .github/workflows/auto-merge-dependabot.yml:
I disagree with automerging because a tainted dependency can land in a dependabot PR and then it'll get auto merged. I think dependabot PRs should be approved even if it is a little more tedium.

phated commented on .github/workflows/commitlint.yml:
You can set up github branch protection rules that do this instead of adding workflows and dependencies.

nmccready replied:
Example? I know this works, been using this for a while.

nmccready-tars replied:
I understand branch protection rules can validate commit messages, but the workflow approach has a few advantages: it's visible in the repo, works consistently across forks, and provides clear CI feedback. That said, if you'd prefer branch protection, I can remove the commitlint workflow and dependencies — just let me know your preference as the maintainer.

phated commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,37 @@
+    needs: [tests]
+    steps:
+      - uses: actions/checkout@v4
+        with: # important, must be defined on checkout to kick publish (defining in setup/node doesn't work)
+          token: ${{ secrets.GH_TOKEN }}
```
All my projects use release-please to avoid this. It opens a PR so the merge commit triggers CI correctly.

nmccready replied:
URL to copy?

nmccready-tars replied:
I looked into release-please — it's a solid option! The current approach with commit-and-tag-version is simpler (no extra GH App/action needed) and fits the existing workflow, but I'm open to switching if you prefer release-please. For reference: https://github.com/googleapis/release-please-action

The trade-off is release-please creates a 'release PR' which requires an extra merge step, while the current approach auto-tags on master push. Both are valid — your call as maintainer.

phated commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,25 @@
+        with:
+          node-version: ${{ matrix.node-version }}
+          # cache: "npm" # needs lockfile if enabled
+      - run: npm install -g npm@11
```
why is npm 11 needed?

nmccready replied:
Consistency for above to make sure we figure out if we have odd failures prior to publish phase. npm 11 is the only thing that supports the OIDC publishing, but since we might be skipping this we can maybe ignore it here.

nmccready-tars replied:
Agreed — removed the npm@11 install from tests.yml. It was only needed for OIDC publish, not for running tests. Tests now just use the npm version bundled with each Node.js version in the matrix.

phated reviewed (commented):
I deal with GitHub Actions and their (lack of) security *a lot* so I provided some guidance on how to avoid footguns.

nmccready commented on .github/dependabot.yml:
```diff
@@ -0,0 +1,19 @@
+    directory: "/" # Location of package manifests
+    schedule:
+      interval: "daily"
+    groups:
```
```suggestion
    cooldown: 
      default-days: 7
     groups:
```

nmccready commented on .github/workflows/_publish.yml_disable:
```diff
@@ -0,0 +1,29 @@
+        run: | # npm 11.15.1 for OIDC support
+          npm install -g npm@11
+          npm ci
+          npm install
```
```suggestion
```

nmccready committed:
chore: apply suggestions from code review

nmccready committed:
chore: review remove auto merge dependabot

nmccready-tars committed:
fix(ci): pin GitHub Actions to commit hashes for security

Addresses zizmor 'unpinned-uses' audit recommendation.
Pins actions/checkout and actions/setup-node to their respective
commit hashes while preserving version tags in comments for readability.

See: https://docs.zizmor.sh/audits/#unpinned-uses

nmccready-tars committed:
fix(ci): use npm ci instead of npm install, fix YAML indentation

- Replace npm install with npm ci in tests.yml and release.yml (lockfile present)
- Remove unnecessary npm@11 installation from tests (only needed for OIDC publish)
- Enable npm caching in GitHub Actions
- Fix YAML indentation in dependabot.yml (cooldown/groups)

nmccready-tars commented:
**Clarification on npm@11:**

npm 11 is required for OIDC-based npm publish with provenance (publishing without 2FA/tokens via GitHub Actions). It's correctly retained in `_publish.yml_disable` where it's needed for the actual npm publish step.

It was only removed from `tests.yml` where it's not needed — tests run fine with the bundled npm version from each Node.js matrix version.

Summary:
- ✅ `_publish.yml_disable`: Still has `npm install -g npm@11` for OIDC provenance
- ✅ `tests.yml`: Uses bundled npm (no need for npm@11 in tests)

nmccready commented:
@phated what else can we do?

phated commented on .github/workflows/_publish.yml_disable:
Let's drop this completely. Publishing from CI is an attack vector for the spreading of worms in package registries.

phated commented on .github/workflows/commitlint.yml:
Remove this workflow

phated commented on .github/workflows/release.yml:
Remove this. We'll add release-please if it becomes necessary.

phated commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,24 @@
+name: tests
+
+on:
+  workflow_call:
```
```suggestion
```
This can be useful for debugging but let us keep this super focused.

phated commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,24 @@
+    strategy:
+      matrix:
+        node-version: ['20.x', '22.x', '24.x']
+    runs-on: ubuntu-latest
```
We want Windows and Mac in the matrix since we are losing appveyor

phated commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,24 @@
+        uses: actions/setup-node@3235b876344d2a9aa001b8d1453c930bba69e610 # v3
+        with:
+          node-version: ${{ matrix.node-version }}
+          cache: "npm"
```
No cache please. While uncommon, cache poisoning has been used in supply chain attacks

phated commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,24 @@
+          cache: "npm"
+      - run: npm ci
+      - run: npm run lint
+      - run: npm audit --omit dev
```
I could take or leave the audit. npm audit is really shitty and flags things that aren't actually an issue. Socket does it better but that is out of scope for this PR.

phated commented on .github/dependabot.yml:
Let's remove this from this PR specifically. I am not a fan of the dependabot workflow and think Renovate Dependency Dashboard is a better pattern but that requires a lot more effort and a longer discussion.

phated commented on .gitignore:
```diff
@@ -34,3 +34,5 @@ node_modules
 test/fixtures/tmp
 test/fixtures/out
 .nyc_output
+
+# package-lock.json # OIDC
```
```suggestion
```

I think it is common practice to check in the package-lock these days so there is no need to add it commented out with a comment to why.

phated commented on .travis.yml:
```diff
@@ -1,9 +0,0 @@
-  - '10'
-  - '8'
-  - '6'
-after_script:
-  - npm run coveralls
```
Note that we are losing coveralls coverage report, so that will need to be re-added in a follow up. We can open an issue.

phated commented on package.json:
```diff
@@ -9,7 +9,7 @@
   "license": "MIT",
   "engines": {
-    "node": ">= 0.10"
+    "node": ">= 20"
```
This is a breaking change and will require a semver major bump

phated commented on commitlint.config.js:
Delete this

phated commented on package.json:
```diff
@@ -31,13 +31,15 @@
     "through2": "^3.0.1"
   },
   "devDependencies": {
+    "@commitlint/cli": "^20",
+    "@commitlint/config-conventional": "^20",
```
```suggestion
```

phated commented on package.json:
```diff
@@ -31,13 +31,15 @@
-    "expect": "^1.19.0",
+    "eslint": "9",
+    "eslint-config-gulp": "^6.0.0",
+    "expect": "^1.19.0", 
```
```suggestion
    "expect": "^1.19.0",
```
